### ISSUE TITLE
Rethink starting questions from "Saved Questions" pseudo-database

### ISSUE THREAD

**Issue by tlrobinson** [metabase/metabase#10163]
Currently we have a fake "Saved Questions" database that includes every saved question as a fake table, which you can start a new question based on. We load it using `GET /api/database?include_tables=true&include_cards=true` to populate the data picker in the query builder. There are a bunch of problems with this:

1. Every saved question appears as a table in that database, which we load when starting the query builder. On large instances this can drastically slow down the first load of the query builder (including on our internal instance, where it takes several seconds to first load the query builder)
2. Since these fake tables belong to the fake "Saved Questions" database rather than their underlying database we don't have the correct database "features" list (joins, etc), so it's impossible to join when starting from a saved question (for example).
3. Similarly, we can only join with saved questions from the same database, but there's no way to know which saved question tables belong to the current database. Currently the UI allows you to join with any saved question, even if it's from a different underlying database, which will result in a failed query (`Failed to fetch Table 1234: Table does not exist, or belongs to a different Database.`)
4. The UX on instances with many questions isn't great because
  a. there are many questions which don't make sense to use as a starting point for a new question (e.x. any scalar query) but we show them anyway
  b. the Saved Questions "database" presents collections as a flat list of schemas, even though we added hierarchical collections
5. IIRC the "Saved Questions" database was a bit of a workaround for not supporting nested queries in the query builder. Since we're adding the ability to do nesting directly in the query builder "notebook" it seems less necessary. You can just open the question you wanted to start from, switch to notebook mode, and continue adding notebook "steps" from there (although we don't currently allow mixing native and MBQL queries) 

I propose the following:

1. Remove the `Saved Questions` database entirely. Power users can simply open any  existing question in notebook mode and continue adding steps, then save as a new question (starting from native queries is the one current caveat).
2. Allow admins to "pin" specific saved questions as new tables that appear alongside real tables (and set permissions on them like real tables). These would be curated tables that easy-mode users would be able to use without knowing the difference between "Saved 
Question" tables and real tables.

**Comment by mazameli** [comment]
Questions:
- Who should be able to do this? Current bias: just admins.
- Where/how exactly in the UI should this be done?
- How will this change the data selection step UI/IX for new questions?
- Should there still be a way to browse through all saved questions beyond the ones that have been marked as being good starting points?

**Comment by antoine-lizee** [comment]
What brought me here is trying to understand why viewing a question got really slow for us recently (making MB much less usable). After checking calls, I found the culprit:
 - `database?include_tables=true&include_cards=true` takes about 2-5 seconds (120+kb of json goodness to serialize, from our 3K `cards`)
 - this calls is fired **systematically*, even to see a question that has been built with the SQL or the simple question viewer
 - was this call added to 0.33 as a result of the new builder?

In our organization, we rely heavily on metabase to quickly share and check results, and people typically open (many) tabs from links shared on slack, etc... The caching that seems to happen in subsequent calls is very little benefit to our usage, while making this call upfront is very expensive.

I understand that this thread is about rethinking the "saved question database", and **we would be heavily in favor of dropping it entirely** (first proposition above), since we consider this feature inapplicable in real life (with teams, people and processes) and overall buggy.

Here, I am proposing a mitigation in the meantime that requires no change or regression:  Can we avoid getting from the backend all the cards until we need them? Especially, remove the call `database?include_tables=true&include_cards=true` for SQL-built or simple questions.

**Comment by tlrobinson** [comment]
@antoine-lizee As far as I can tell, the only thing that changed between 0.32.x and 0.33.x with respect to `/api/database?include_tables=true&include_cards=true` is that we no longer start loading it on the `/question/new` page, which may have made it appear faster when starting a new question. We also cache the result so it only happens once.

I thought that perhaps previously we loaded the `/api/card` and `/api/card/query` in parallel with that `/api/database` request, but it doesn't look like that's the case (this screenshot is with intentional 5s latency):

<img width="997" alt="Screenshot 2019-09-19 15 31 43" src="https://user-images.githubusercontent.com/18193/65285931-be3b2a00-daf2-11e9-9100-5df6375e9ef0.png">

We could revert to the previous behavior of loading it when opening the "new question" page, but that wouldn't help with saved questions. We could just load it up front when you start the app. Or we could load it in parallel. We could also avoid `include_cards=true` for SQL queries.

**Comment by tlrobinson** [comment]
@antoine-lizee I think #10948 should improve the situation, if you'd like to take a look.

**Comment by tlrobinson** [comment]
#10948 unblocks loading the saved query.

Technically it still loads `/api/database?include_tables=true&include_cards=true` in all variants of the query builder (new / saved queries, MBQL/native, view/notebook) but the user shouldn't notice that, and in most cases it will improve perceived latency if they do go change the table (not withstanding putting extra load on the backend)

A couple other ideas that wouldn't require the bigger changes described in the original issue:

1. Leave off `include_cards=true` for SQL queries (but not `include_tables=true` because some dbs like MongoDB actually require a table for native queries).
2. Split it up into two requests, `/api/database?include_tables=true` and `/api/database?include_cards=true` and only load the ones we think will be required.
3. Skip preloading it up front in some cases (something like `loadTables = isNew || mode === "notebook"`, `loadCards = (isNew || mode === "notebook") && !isNative`)
4. Load each database (including "saved questions" fake db) on demand as the user opens different sections in the data picker.

We could implement any combination of these. It's all tradeoffs between caching requests, backend/db load, user-perceived latency.

**Comment by antoine-lizee** [comment]
@tlrobinson thanks for the PR, I [commented](https://github.com/metabase/metabase/pull/10948#issuecomment-535093503) on it directly.

Ideas, starting from yours:
1. -> agreed, this was I am suggesting in the PR. would solve the pb for SQL at least.
2. Yes! But we will need them for the query builders anyway
4. will incur not-so-great UX, of the type that we're seeing when exploring the data model in the admin panel (really awful to wait for 3 seconds before the database appears after selection)

5. I believe the "best" solution if we keep the saved questions would be to have the best of both world
  - the 2 queries be eagerly preloaded
  - don't block anything in the UI _until we need it_. This is especially true for the saved questions.
-> this will not be straighforward to implement but have a lot of long-term value if we want to keep the ability to build from saved questions

6. **Get rid of the saved questions** (number of tables << number of saved questions typically).

**Comment by flamber** [comment]
Closing this, since the way Saved Questions are loaded has completely changed in 0.40
